### PR TITLE
CVSL-1847 inactivate timed out licence on activation

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceExpiryIntegrationTest.kt [hunk 1]
```diff
@@ -49,6 +49,7 @@ class LicenceExpiryIntegrationTest : IntegrationTestBase() {
         Tuple.tuple(2L, LicenceStatus.INACTIVE),
         Tuple.tuple(5L, LicenceStatus.INACTIVE),
         Tuple.tuple(6L, LicenceStatus.INACTIVE),
+        Tuple.tuple(7L, LicenceStatus.INACTIVE),
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceExpiryIntegrationTest.kt:52]: new timed out licence

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceActivationIntegrationTest.kt [hunk 2]
```diff
@@ -80,14 +88,15 @@ class LicenceActivationIntegrationTest : IntegrationTestBase() {
...
-    assertThat(deactivatedLicences?.size).isEqualTo(2)
+    assertThat(deactivatedLicences?.size).isEqualTo(3)
     assertThat(deactivatedLicences)
       .extracting<Tuple> {
         tuple(it.licenceId, it.licenceStatus)
       }
       .contains(
         tuple(4L, LicenceStatus.INACTIVE),
         tuple(5L, LicenceStatus.INACTIVE),
+        tuple(6L, LicenceStatus.INACTIVE),
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceActivationIntegrationTest.kt:99]: a new timed out licence

File: src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceActivationIntegrationTest.kt [hunk 3]
```diff
@@ -69,6 +76,7 @@ class LicenceActivationIntegrationTest : IntegrationTestBase() {
         tuple(1L, LicenceStatus.ACTIVE),
         tuple(2L, LicenceStatus.ACTIVE),
         tuple(3L, LicenceStatus.ACTIVE),
+        tuple(7L, LicenceStatus.ACTIVE),
```
Comment (andrewrlee) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/integration/LicenceActivationIntegrationTest.kt:79]: a new activation for an approved hard stop licence

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt [hunk 4]
```diff
@@ -110,10 +110,10 @@ class LicenceCreationService(
     val createdBy = staffRepository.findByUsernameIgnoreCase(username) as PrisonUser?
       ?: error("Staff with username $username not found")
 
-    val timedOutLicence: CrdLicence? = licenceRepository.findByBookingIdAndStatusCodeOrderByDateCreatedDesc(
-      nomisRecord.bookingId!!.toLong(),
+    val timedOutLicence: CrdLicence? = licenceRepository.findAllByBookingIdInAndStatusCodeOrderByDateCreatedDesc(
+      listOf(nomisRecord.bookingId!!.toLong()),
       TIMED_OUT,
-    )
+    ).firstOrNull()
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceCreationService.kt:116]: just adapting to method signature change

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt [hunk 5]
```diff
@@ -194,6 +196,10 @@ class LicenceService(
           listOf(licenceEntity),
           "Deactivating licence as the parent licence version was activated",
         )
+        inactivateTimedOutLicenceVersions(
+          listOf(licenceEntity),
+          "Deactivating licence as the parent licence version was activated",
+        )
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt:202]: inactivate time stopped licences when changed via support screen

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt [hunk 6]
```diff
@@ -488,6 +494,7 @@ class LicenceService(
         activatedLicences,
         "Licence automatically deactivated as the approved licence version was activated",
       )
+      inactivateTimedOutLicenceVersions(activatedLicences, "Licence automatically deactivated as the approved licence version was activated")
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceService.kt:497]: Called as part of the activation job for IS-91 licences and other licences where the release event failed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceOverrideService.kt [hunk 7]
```diff
@@ -66,6 +67,10 @@ class LicenceOverrideService(
 
     if (newStatus == LicenceStatus.ACTIVE) {
       licenceActivatedDate = LocalDateTime.now()
+      licenceService.inactivateTimedOutLicenceVersions(
+        listOf(licence),
+        "Deactivating timed out licence as the licence status was overridden to active",
+      )
```
Comment (andrewrlee) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/LicenceOverrideService.kt:73]: https://github.com/ministryofjustice/create-and-vary-a-licence-api/pull/515

**Review by gurps-singh — APPROVED** [review]

**Pull request merged**